src/capi/typeobject.cpp

```diff
@@ -3316,8 +3316,24 @@ extern "C" int PyType_Ready(PyTypeObject* cls) noexcept {
     cls->gc_visit = &conservativeGCHandler;
     cls->is_user_defined = true;
 
-    // this should get automatically initialized to 0 on this path:
-    assert(cls->attrs_offset == 0);
+
+    bool is_metaclass = PyType_IsSubtype(cls, type_cls);
+    if (is_metaclass) {
+        assert(!cls->instancesHaveHCAttrs() && !cls->instancesHaveDictAttrs() && cls->tp_base);
+        if (cls->tp_base->instancesHaveHCAttrs()) {
+            // This doesn't get copied in inherit_slots like other slots do.
+            cls->attrs_offset = cls->tp_base->attrs_offset;
+        }
+
+        // If this class is a metaclass defined in a C extension, chances are that some of its
+        // instances may be hardcoded in the C extension as well. Those instances will call
+        // PyType_Ready and expect their class (this metaclass) to have a place to put attributes.
+        // e.g. CTypes does this.
+        assert(cls->instancesHaveHCAttrs() || cls->instancesHaveDictAttrs());
+    } else {
+        // this should get automatically initialized to 0 on this path:
+        assert(cls->attrs_offset == 0);
+    }
```

Comment:
Maybe we should just always inherit attrs_offset?  I think this will affect any case that an extension class inherits from a Pyston/Python class.